tests/sgx/wrfsbase/enc/enc.c

```diff
@@ -8,34 +8,132 @@
+    // than 2 bytes.
+    register int edi __asm__("edi") = 0;
+    asm volatile("idiv %1"
+                 : "=a"(ret)
+                 : "r"(edi) // Divisor of 0 is hard-coded
+                 : "%1",
+                   "cc"); // cc indicates that flags will be clobbered by ASM
+
+    // Check if the float registers are recovered correctly after the exception
+    // is handled.
+    if (f < 0.309 || f > 0.321 || d < 0.319 || d > 0.321)
+    {
+        return -1;
+    }
+
+    return 0;
+}
+
+static uint64_t _divide_by_zero_handler(oe_exception_record_t* exception_record)
+{
+    void* current_fs;
+
+    asm volatile("mov %%fs:0, %0" : "=r"(current_fs));
+
+    OE_TEST(current_fs == original_fs);
+
+    if (exception_record->code != OE_EXCEPTION_DIVIDE_BY_ZERO)
+        return OE_EXCEPTION_CONTINUE_SEARCH;
+
+    // Skip the idiv instruction - 2 is tied to the size of the idiv instruction
+    // and can change with a different compiler/build. Minimizing this with the
+    // use of the inline assembly for integer division
+    exception_record->context->rip += 2;
+    return OE_EXCEPTION_CONTINUE_EXECUTION;
+}
+
 void enc_wrfsbase(int negative_test)
 {
     static uint64_t temp_td[256];
-    void* old_fs;
-    void* new_fs;
+    void* fs_after_exception;
+    void* fs_after_cpuid;
     void* recovered_fs;
+    void* new_fs;
+    void* new_fs_2;
+
+    OE_TEST(
+        oe_add_vectored_exception_handler(true, _divide_by_zero_handler) ==
+        OE_OK);
 
     temp_td[0] = (uint64_t)temp_td;
 
-    asm volatile("mov %%fs:0, %0" : "=r"(old_fs));
+    asm volatile("mov %%fs:0, %0" : "=r"(original_fs));
 
     /* change FS */
     asm volatile("wrfsbase %0 " : : "a"(temp_td));
     asm volatile("mov %%fs:0, %0" : "=r"(new_fs));
 
+    /* Expect FS is persisted after CPUID emulation */
+    {
+        uint32_t cpuid_rax = 0;
+        uint32_t ebx = 0;
+        uint32_t ecx = 0;
+        uint32_t edx = 0;
+
+        _cpuid(0, 0, &cpuid_rax, &ebx, &ecx, &edx);
+    }
+
+    asm volatile("mov %%fs:0, %0" : "=r"(fs_after_cpuid));
+
+    /* Expect FS is restored after an exception */
+    OE_TEST(_divide_by_zero_exception_function() == 0);
+
+    asm volatile("mov %%fs:0, %0" : "=r"(fs_after_exception));
+
+    /* change FS */
+    asm volatile("wrfsbase %0 " : : "a"(temp_td));
+    asm volatile("mov %%fs:0, %0" : "=r"(new_fs_2));
+
     if (negative_test)
     {
-        oe_sgx_td_t* td = oe_sgx_get_td();
-        /* unreachable, dummy test to use td */
-        OE_TEST(td->state == OE_TD_STATE_ABORTED);
+        /* Calling OCALLs will fail if FS is changed */
+        host_dummy();
     }
 
     /* restore FS */
-    asm volatile("wrfsbase %0 " : : "a"(old_fs));
+    asm volatile("wrfsbase %0 " : : "a"(original_fs));
     asm volatile("mov %%fs:0, %0" : "=r"(recovered_fs));
 
     OE_TEST(new_fs == temp_td);
-    OE_TEST(recovered_fs == old_fs);
+    OE_TEST(fs_after_cpuid == temp_td);
+    OE_TEST(fs_after_exception == original_fs);
```

Comment:
It is counter-intuitive that when control resumes after an exception, the FS value is not the value just before exception.

Reply:
updated.